inst/include/dust/densities.hpp

```diff
@@ -91,11 +90,24 @@ HOSTDEVICE T dnbinom(int x, T size, T mu, bool log) {
   if (mu == 0) {
     return maybe_log(x == 0 ? 0 : -dust::utils::infinity<T>(), log);
   }
-  const T ret = dust::utils::lgamma(static_cast<T>(x + size)) -
-    dust::utils::lgamma(static_cast<T>(size)) -
-    dust::utils::lgamma(static_cast<T>(x + 1)) +
-    size * std::log(prob) + x * std::log(1 - prob);
-  return maybe_log(ret, log);
+
+  // Avoid size / (size + mu) where this would cause prob to be
+  // equal. Somewhat arbitrarily, taking 100 * floating point eps as
+  // the change over.
+  const T ratio = dust::utils::epsilon<T>() * 100;
+  if (mu < ratio * size) {
+    const T p = std::log(mu / (1 + mu / size));
+    const T ret = x * p - mu - lgamma(x + 1) +
+      std::log1p(x * (x - 1) / (2 * size));
+    return maybe_log(ret, log);
```

Comment:
```suggestion
    const T log_prob = std::log(mu / (1 + mu / size));
    const T ret = x * log_prob - mu - lgamma(x + 1) +
      std::log1p(x * (x - 1) / (2 * size));
    return maybe_log(ret, log);
```

Reply:
Where does this formulation come from, I couldn't see how the `ret = ...` formula follows from size =~ 0